### PR TITLE
Immediately apply deferred system params in System::run

### LINKED ISSUE #11821: `System::run` should call `System::apply_deferred`
## What problem does this solve or what need does it fill?

Since merging #9822, we're now treating deferred mutations the same way as other mutations when possible. We don't want users to have to think about when commands get applied, so we apply them as early as possible in most cases. 

## What solution would you like?

In the default implementation for `System::run`, we should call `System::apply_deferred` after calling `System::run_unsafe`. We already have access to `&mut World` in `System::run`, so there is little reason not to do this.

## What alternative(s) have you considered?

None

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Fixes #11821. 

## Solution
* Run `System::apply_deferred` in `System::run` after executing the system.
* Switch to using `System::run_unsafe` in `SingleThreadedExecutor` to preserve the current behavior.
* Remove the `System::apply_deferred` in `SimpleExecutor` as it's now redundant.
* Remove the `System::apply_deferred` when running one-shot systems, as it's now redundant. 

---

## Changelog
Changed: `System::run` will now immediately apply deferred system params after running the system.

## Migration Guide
`System::run` will now always run `System::apply_deferred` immediately after running the system now. If you were running systems and then applying their deferred buffers at a later point in time, you can eliminate the latter.

```rust
// in 0.13
system.run(world);
// .. sometime later ...
system.apply_deferred(world);

// in 0.14
system.run(world);
```

### CODE REVIEW

File: crates/bevy_ecs/src/system/system.rs [hunk 1]
```diff
@@ -66,14 +66,20 @@ pub trait System: Send + Sync + 'static {
     ///
     /// For [read-only](ReadOnlySystem) systems, see [`run_readonly`], which can be called using `&World`.
     ///
+    /// Unlike [`System::run_unsafe`], this will apply deferred parameters *immediately*.
+    ///
     /// [`run_readonly`]: ReadOnlySystem::run_readonly
     fn run(&mut self, input: Self::In, world: &mut World) -> Self::Out {
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/system/system.rs:72]: The docs for `run_unsafe` should also clarify this distinction

File: crates/bevy_ecs/src/system/system.rs [hunk 2]
```diff
@@ -66,14 +66,20 @@ pub trait System: Send + Sync + 'static {
...
-        // - `update_archetype_component_access` has been called.
-        unsafe { self.run_unsafe(input, world) }
+        let ret = {
+            let world = world.as_unsafe_world_cell();
+            self.update_archetype_component_access(world);
+            // SAFETY:
+            // - We have exclusive access to the entire world.
+            // - `update_archetype_component_access` has been called.
+            unsafe { self.run_unsafe(input, world) }
+        };
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/system/system.rs:80]: Is this block necessary?

File: crates/bevy_ecs/src/schedule/executor/single_threaded.rs [hunk 3]
```diff
@@ -96,16 +96,26 @@ impl SystemExecutor for SingleThreadedExecutor {
...
+            let res = std::panic::catch_unwind(AssertUnwindSafe(|| {
+                if system.is_exclusive() {
                     system.run((), world);
-                }));
-                if let Err(payload) = res {
-                    eprintln!("Encountered a panic in system `{}`!", &*system.name());
-                    std::panic::resume_unwind(payload);
+                } else {
+                    // Use run_unsafe to avoid immediately applying deferred buffers
+                    let world = world.as_unsafe_world_cell();
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/executor/single_threaded.rs:107]: I'd rather we just make `single_threaded` executor work the same way as `simple` executor. We should make it clear that `apply_deferred` is just an optimization to allow systems to run in parallel. Any single-threaded executor should apply commands immediately -- it has access to `&mut World` so there's little reason not to.
Although I suppose deprecating `SimpleExecutor` can be for another PR.

**Review by JoJoJet — APPROVED** [review]

File: crates/bevy_ecs/src/system/system.rs (hunk 2 above)
Comment (james7132) [reply crates/bevy_ecs/src/system/system.rs:80]: `apply_deferred` requires a `&mut World`, and these require `UnsafeWorldCell`. The block might be avoidable if we don't shadow the variable name, but I'm still not sure about the scoping for that.

Comment (JoJoJet) [reply crates/bevy_ecs/src/system/system.rs:80]: I'd prefer we just not shadow it. The scoping should be fine, unsafe-world-cell is usually well-behaved as a reference type.

File: crates/bevy_ecs/src/schedule/executor/simple.rs [hunk 4]
```diff
@@ -36,73 +35,11 @@ impl SystemExecutor for SimpleExecutor {
         world: &mut World,
         _skip_systems: Option<&FixedBitSet>,
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/executor/simple.rs:36]: Nit: since we're actually using this argument, I think that we should remove the _ from the name.

File: crates/bevy_ecs/src/schedule/executor/single_threaded.rs [hunk 5]
```diff
@@ -95,28 +90,24 @@ impl SystemExecutor for SingleThreadedExecutor {
...
         }
+
         self.evaluated_sets.clear();
         self.completed_systems.clear();
     }
 
-    fn set_apply_final_deferred(&mut self, apply_final_deferred: bool) {
-        self.apply_final_deferred = apply_final_deferred;
+    fn set_apply_final_deferred(&mut self, _: bool) {
+        // do nothing. single threaded executor does not do a final sync
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/executor/single_threaded.rs:110]:
```suggestion
        // do nothing. single threaded executor does not do a final sync
        // as commands are always immediately applied
```

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Much simpler. Feel free to merge once you get CI passing.

**Comment by hymm** [comment]
Is this going to mess up pipelined rendering extract schedule if you run all the schedules single threaded?

**Comment by james7132** [comment]
That would likely be a performance regression. That said I'm not sure why one would be running pipelined rendering with a single threaded executor. It seems counter intuitive.

**Comment by james7132** [comment]
@JoJoJet I think that's good enough justification to not deprecate the SimpleExecutor and retain the behavior in the single threaded executor. Thoughts?
